Add RandID to BatchSpecExecutions and use it resolver

mrnugget opened the pull request:
Part of https://github.com/sourcegraph/sourcegraph/issues/22550.

This brings `batch_spec_executions` up to date with the pattern we use for `batch_specs` and `changeset_specs` too.

Note: we need to delete all batch_spec_executions before running this migration since I added a non-null constraint. Should be fine since only Erik and I have those on our machine. But we need to delete them on k8s.sgdev.org before we merge this.

eseliger commented on migrations/frontend/1528395848_add_rand_id_to_batch_spec_executions.down.sql:
```diff
@@ -0,0 +1,8 @@
+
```
```suggestion
```

eseliger commented on migrations/frontend/1528395848_add_rand_id_to_batch_spec_executions.up.sql:
```diff
@@ -0,0 +1,7 @@
+
+ALTER TABLE IF EXISTS batch_spec_executions ADD COLUMN IF NOT EXISTS rand_id text NOT NULL;
+
+CREATE INDEX batch_spec_executions_rand_id ON batch_spec_executions USING btree (rand_id);
```
not unique?

mrnugget replied:
We don't have a unique constraint on it since we don't handle that failing in the application yet. See: https://github.com/sourcegraph/sourcegraph/issues/16721 I guess we can update that ticket now :grimacing:

eseliger commented on enterprise/internal/batches/store/batch_spec_execution.go:
```diff
@@ -112,8 +124,14 @@ LIMIT 1
 func getBatchSpecExecutionQuery(opts *GetBatchSpecExecutionOpts) *sqlf.Query {
-	preds := []*sqlf.Query{
-		sqlf.Sprintf("id = %s", opts.ID),
+	var preds []*sqlf.Query
+
+	if opts.ID != 0 {
+		preds = append(preds, sqlf.Sprintf("id = %s", opts.ID))
+	}
+
+	if opts.RandID != "" {
+		preds = append(preds, sqlf.Sprintf("rand_id = %s", opts.RandID))
 	}
```
if neither are set, it would return _any_ row, do we think that's dangerous?

mrnugget replied:
I added a check for empty `preds` and return an error now. Good call.

eseliger reviewed (approved):
Nice 🌟 For the other entities that have this, we added a quite extensive doc block to the graphql schema about how it's safe and why that is needed. Do you think we should add that too?

mrnugget commented:
> we added a quite extensive doc block to the graphql schema about how it's safe and why that is needed. Do you think we should add that too?

Added that!

courier-new reviewed (approved):
Looks good to me! It's cool to see this in the context of adding a new `RandID` to some entity and what all that entails.
